Project 4

Import Modules

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

Next I will load the mushroom data

In [3]:
url = "https://raw.githubusercontent.com/dwaynewebb86/IS362_Assignment8/main/agaricus-lepiota.data"
df = pd.read_csv(url, header=None)

Now I will assign column names based on the information provided in the names file and print the head for investigation

In [4]:
df.columns = [
    "class","cap-shape","cap-surface","cap-color","bruises","odor",
    "gill-attachment","gill-spacing","gill-size","gill-color",
    "stalk-shape","stalk-root","stalk-surface-above-ring",
    "stalk-surface-below-ring","stalk-color-above-ring",
    "stalk-color-below-ring","veil-type","veil-color","ring-number",
    "ring-type","spore-print-color","population","habitat"
]

df.head()

,class,cap-shape,cap-surface,cap-color,bruises,odor,gill-attachment,gill-spacing,gill-size,gill-color,...,stalk-surface-below-ring,stalk-color-above-ring,stalk-color-below-ring,veil-type,veil-color,ring-number,ring-type,spore-print-color,population,habitat
0,p,x,s,n,t,p,f,c,n,k,...,s,w,w,p,w,o,p,k,s,u
1,e,x,s,y,t,a,f,c,b,k,...,s,w,w,p,w,o,p,n,n,g
2,e,b,s,w,t,l,f,c,b,n,...,s,w,w,p,w,o,p,n,n,m
3,p,x,y,w,t,p,f,c,n,n,...,s,w,w,p,w,o,p,k,s,u
4,e,x,s,g,f,n,f,w,b,k,...,s,w,w,p,w,o,e,n,a,g


Now I will convert the variables

In [5]:
df['class'] = df['class'].map({'p':1, 'e':0})

Now I will select the predictors of odor and spore print-color, and the target variable of class.

In [6]:
X_odor = pd.get_dummies(df[['odor']])
X_spore = pd.get_dummies(df[['spore-print-color']])
y = df['class']


Now I will split the data into training and testing sets using a 70/30 split. This allows me to train the model and then evaluate its performance on unseen data. I performed the split separately for odor and spore print color, and set random_state=42 to keep the results consistent.

In [7]:
Xo_train, Xo_test, yo_train, yo_test = train_test_split(X_odor, y, test_size=0.3, random_state=42)
Xs_train, Xs_test, ys_train, ys_test = train_test_split(X_spore, y, test_size=0.3, random_state=42)

Next I trained a Decision Tree model using the odor data. I then used the model to make predictions on the test set and calculated the accuracy to evaluate how well odor predicts whether a mushroom is poisonous.

In [8]:
model_odor = DecisionTreeClassifier()
model_odor.fit(Xo_train, yo_train)

pred_odor = model_odor.predict(Xo_test)

odor_accuracy = accuracy_score(yo_test, pred_odor)
print("Odor Accuracy:", odor_accuracy)

Odor Accuracy: 0.9840032813781788


Lastly I trained a Decision Tree model using the spore print color data. I then used the model to make predictions on the test set and calculated the accuracy to evaluate how well this feature predicts whether a mushroom is poisonous.

In [9]:
model_spore = DecisionTreeClassifier()
model_spore.fit(Xs_train, ys_train)

pred_spore = model_spore.predict(Xs_test)

spore_accuracy = accuracy_score(ys_test, pred_spore)
print("Spore Print Color Accuracy:", spore_accuracy)

Spore Print Color Accuracy: 0.8695652173913043


Based on the analysis, I can see that odor has a very strong relationship with whether a mushroom is edible or poisonous. The model using odor produced an accuracy of 0.9840, which indicates that certain odor values are closely associated with poisonous mushrooms. This suggests that odor is a highly useful feature for predicting the class of a mushroom and can almost fully separate edible from poisonous cases on its own.
When looking at spore print color, I do not see as strong of a relationship with the class variable. The model produced an accuracy of 0.8696, which shows that it does have some predictive value, but it is not as strong as odor. Both edible and poisonous mushrooms appear across multiple spore print color values without a clear separation. This suggests that spore print color may provide some predictive value, but it is not as strong of a predictor when used by itself.
Overall, odor stands out as the better predictor between the two variables tested, while spore print color appears to be more useful when combined with other features rather than used independently.